[JSC] Async Function should be driven directly via MicrotaskQueue

Constellation opened the pull request:
#### f70f597b849ebd6561da99753d8e81d7586650d3
<pre>
[JSC] Async Function should be driven directly via MicrotaskQueue
<a href="https://bugs.webkit.org/show_bug.cgi?id=302677">https://bugs.webkit.org/show_bug.cgi?id=302677</a>
<a href="https://rdar.apple.com/164925036">rdar://164925036</a>

Reviewed by Yijia Huang.

We were scheduling next invocation of async function via
@resolveWithoutPromiseForAsyncAwait and passing
@asyncFunctionResumeOnFulfilled and @asyncFunctionResumeOnRejected
functions. And @asyncFunctionResume is invoked from them. But this is
inefficient in two ways.

1. These functions are completely builtin. We do not need to have them
   in JS.
2. MicrotaskQueue invokes these functions from C++ world, and then
   @asyncFunctionResume invokes actual async function in the one place.
   This is double-dispatch and we do not need to pay this cost. It
   should be only one dynamic dispatch.

This patch implements these builtin JS logic in MicrotaskQueue itself.
And now MicrotaskQueue directly dispatches user-defined async functions.
JSPromise / MicrotaskQueue extend WithInternalMicrotask variants and
start using it for these async function driving. In the future, we would
like to completely replace WithoutPromise variants with this WithInternalMicrotask
variants.

We also inline asyncFunctionResume completely to the async wrapper
function itself. This function is doing double-dispatch unnecessary and
we should have unique call site into the user-provided function not to
cause megamorphic calls.

We also shrink the size of JSGenerator and JSAsyncGenerator as we are
not using these PolyProto field so far.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources-input.xcfilelist:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/builtins/AsyncFunctionPrototype.js:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::FunctionNode::emitBytecode):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::getStackTrace):
* Source/JavaScriptCore/runtime/JSAsyncGenerator.h:
* Source/JavaScriptCore/runtime/JSGenerator.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::promiseResolveThenableJobWithInternalMicrotaskFastSlow):
(JSC::runInternalMicrotask):
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSPromise::triggerPromiseReactions):
(JSC::JSPromise::resolveWithInternalMicrotaskForAsyncAwait):
(JSC::JSPromise::resolveWithInternalMicrotask):
(JSC::JSPromise::rejectWithInternalMicrotask):
(JSC::JSPromise::fulfillWithInternalMicrotask):
* Source/JavaScriptCore/runtime/JSPromise.h:
* Source/JavaScriptCore/runtime/Microtask.h:

Canonical link: <a href="https://commits.webkit.org/303208@main">https://commits.webkit.org/303208@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/45b9e94300a20bddc953041bfb078aae90a51438

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131706 "1 style error") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/4198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/42706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/139215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/4124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/3960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/139215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134652 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/155/builds/4124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/138/builds/42706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/139215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/155/builds/4124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/138/builds/42706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/82437 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/123769 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/155/builds/4124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/42706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/141861 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/130213 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/3867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/42706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/141861 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/3948 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/3960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/141861 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/138/builds/42706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/57077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20476 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/3920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/42706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/163178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/3752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/67368 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/163178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/4012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/3881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/runtime/JSGenerator.h:
```diff
@@ -95,6 +84,36 @@ class JSGenerator final : public JSInternalFieldObjectImpl<6> {
     static JSGenerator* create(VM&, Structure*);
     static Structure* createStructure(VM&, JSGlobalObject*, JSValue);
 
+    int32_t state()
```
Add `const`?

Constellation replied:
Done.

hyjorc1 commented on Source/JavaScriptCore/runtime/JSGenerator.h:
```diff
@@ -95,6 +84,36 @@ class JSGenerator final : public JSInternalFieldObjectImpl<6> {
+        Base::internalField(static_cast<unsigned>(Field::State)).setWithoutWriteBarrier(jsNumber(state));
+    }
+
+    JSValue next()
```
Add `const`?

Constellation replied:
Done.

hyjorc1 commented on Source/JavaScriptCore/runtime/JSGenerator.h:
```diff
@@ -95,6 +84,36 @@ class JSGenerator final : public JSInternalFieldObjectImpl<6> {
+        return Base::internalField(static_cast<unsigned>(Field::Next)).get();
+    }
+
+    JSValue thisValue()
```
Add `const`?

Constellation replied:
Done.

hyjorc1 commented on Source/JavaScriptCore/runtime/JSGenerator.h:
```diff
@@ -95,6 +84,36 @@ class JSGenerator final : public JSInternalFieldObjectImpl<6> {
+        return Base::internalField(static_cast<unsigned>(Field::This)).get();
+    }
+
+    JSValue frame()
```
Add `const`?

Constellation replied:
Done.

hyjorc1 commented on Source/JavaScriptCore/runtime/JSGenerator.h:
```diff
@@ -95,6 +84,36 @@ class JSGenerator final : public JSInternalFieldObjectImpl<6> {
+        return Base::internalField(static_cast<unsigned>(Field::Frame)).get();
+    }
+
+    JSValue context()
```
Add `const`?

Constellation replied:
Done.

hyjorc1 reviewed (approved):
r=me if EWS bots are happy.

Constellation committed:
[JSC] Async Function should be driven directly via MicrotaskQueue
https://bugs.webkit.org/show_bug.cgi?id=302677
rdar://164925036

Reviewed by Yijia Huang.

We were scheduling next invocation of async function via
@resolveWithoutPromiseForAsyncAwait and passing
@asyncFunctionResumeOnFulfilled and @asyncFunctionResumeOnRejected
functions. And @asyncFunctionResume is invoked from them. But this is
inefficient in two ways.

1. These functions are completely builtin. We do not need to have them
   in JS.
2. MicrotaskQueue invokes these functions from C++ world, and then
   @asyncFunctionResume invokes actual async function in the one place.
   This is double-dispatch and we do not need to pay this cost. It
   should be only one dynamic dispatch.

This patch implements these builtin JS logic in MicrotaskQueue itself.
And now MicrotaskQueue directly dispatches user-defined async functions.
JSPromise / MicrotaskQueue extend WithInternalMicrotask variants and
start using it for these async function driving. In the future, we would
like to completely replace WithoutPromise variants with this WithInternalMicrotask
variants.

We also inline asyncFunctionResume completely to the async wrapper
function itself. This function is doing double-dispatch unnecessary and
we should have unique call site into the user-provided function not to
cause megamorphic calls.

We also shrink the size of JSGenerator and JSAsyncGenerator as we are
not using these PolyProto field so far.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources-input.xcfilelist:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/builtins/AsyncFunctionPrototype.js:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::FunctionNode::emitBytecode):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::getStackTrace):
* Source/JavaScriptCore/runtime/JSAsyncGenerator.h:
* Source/JavaScriptCore/runtime/JSGenerator.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::promiseResolveThenableJobWithInternalMicrotaskFastSlow):
(JSC::runInternalMicrotask):
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSPromise::triggerPromiseReactions):
(JSC::JSPromise::resolveWithInternalMicrotaskForAsyncAwait):
(JSC::JSPromise::resolveWithInternalMicrotask):
(JSC::JSPromise::rejectWithInternalMicrotask):
(JSC::JSPromise::fulfillWithInternalMicrotask):
* Source/JavaScriptCore/runtime/JSPromise.h:
* Source/JavaScriptCore/runtime/Microtask.h:

Canonical link: https://commits.webkit.org/303208@main

kmiller68 commented on Source/JavaScriptCore/runtime/JSMicrotask.cpp:
```diff
@@ -107,6 +108,35 @@ static void promiseResolveThenableJobWithoutPromiseFastSlow(JSGlobalObject* glob
+    ASSERT(!arguments.hasOverflowed());
+    auto callData = JSC::getCallDataInline(reject);
+    call(globalObject, reject, callData, jsUndefined(), arguments);
+    EXCEPTION_ASSERT(scope.exception() || true);
```
Seems wrong. Why not release the scope instead
